cranelift/codegen/src/isa/x64/inst/mod.rs

```diff
@@ -0,0 +1,974 @@
+    MovZX_M_R {
+        extMode: ExtMode,
+        addr: Addr,
+        dst: Reg,
+    },
+
+    /// A plain 64-bit integer load, since MovZX_M_R can't represent that
+    Mov64_M_R { addr: Addr, dst: Reg },
+
+    /// movs (bl bq wl wq lq) addr reg (good for all SX loads)
+    MovSX_M_R {
+        extMode: ExtMode,
+        addr: Addr,
+        dst: Reg,
+    },
+
+    /// mov (b w l q) reg addr (good for all integer stores)
+    Mov_R_M {
+        size: u8, // 1, 2, 4 or 8
+        src: Reg,
+        addr: Addr,
+    },
+
+    /// (shl shr sar) (l q) imm reg
+    Shift_R {
+        is_64: bool,
+        kind: ShiftKind,
+        nBits: u8, // 1 .. #bits-in-type - 1, or 0 to mean "%cl"
+        dst: Reg,
+    },
+
+    /// cmp (b w l q) (reg addr imm) reg
+    Cmp_RMI_R {
+        size: u8, // 1, 2, 4 or 8
+        src: RMI,
+        dst: Reg,
+    },
+
+    /// pushq (reg addr imm)
+    Push64 { src: RMI },
+
+    /// popq reg
+    Pop64 { dst: Reg },
+
+    /// call simm32
+    CallKnown {
+        dest: ExternalName,
+        uses: Set<Reg>,
+        defs: Set<Writable<Reg>>,
+    },
+
+    /// callq (reg mem)
+    CallUnknown {
+        dest: RM,
+        //uses: Set<Reg>,
+        //defs: Set<Writable<Reg>>,
+    },
+
+    // ---- branches (exactly one must appear at end of BB) ----
+    /// ret
+    Ret {},
+
+    /// A placeholder instruction, generating no code, meaning that a function epilogue must be
+    /// inserted there.
+    EpiloguePlaceholder {},
+
+    /// jmp simm32
+    JmpKnown { dest: BranchTarget },
+
+    /// jcond cond target target
+    // Symmetrical two-way conditional branch.
+    // Should never reach the emitter.
+    JmpCondSymm {
+        cc: CC,
+        taken: BranchTarget,
+        not_taken: BranchTarget,
+    },
+
+    /// Lowered conditional branch: contains the original instruction, and a
+    /// flag indicating whether to invert the taken-condition or not. Only one
+    /// BranchTarget is retained, and the other is implicitly the next
+    /// instruction, given the final basic-block layout.
+    JmpCond {
+        cc: CC,
+        //inverted: bool, is this needed?
+        target: BranchTarget,
+    },
+
+    /// As for `CondBrLowered`, but represents a condbr/uncond-br sequence (two
+    /// actual machine instructions). Needed when the final block layout implies
+    /// that neither arm of a conditional branch targets the fallthrough block.
+    // Should never reach the emitter
+    JmpCondCompound {
+        cc: CC,
+        taken: BranchTarget,
+        not_taken: BranchTarget,
+    },
+
+    /// jmpq (reg mem)
+    JmpUnknown { target: RM },
+}
+
+// Handy constructors for Insts.
+
+// For various sizes, will some number of lowest bits sign extend to be the
+// same as the whole value?
+pub(crate) fn low32willSXto64(x: u64) -> bool {
+    let xs = x as i64;
+    xs == ((xs << 32) >> 32)
+}
+
+impl Inst {
+    pub(crate) fn nop(len: u8) -> Self {
+        debug_assert!(len <= 16);
```

Comment:
Why? You can just emit multiple nop instructions, right?

Reply:
Multiple single-byte nop insns, you mean?  Sure, you *can* do that, but that wastes front end resources.  Intel defines "official" no-ops with lengths at least 1 to 14 and maybe more (I forget).

Reply:
Agreed with Julian, there's an official selection of nop instructions for different sizes.